events: Update extensible events types

zecakeh opened the pull request:
According to the latest rewrite of the MSCs.

The unstable prefixes are now used everywhere, as recommended in MSC1767.

The mixed types have been removed (per the MSC) and the conversion methods too.

The naming of the types has been streamlined by using the `ContentBlock` suffix consistently when a type is defined as such. That also means that some of those types are newtype structs purely for the semantic meaning as a reusable block.

It is recommended to review by file as it usually means reviewing by MSC. It might be easier than by commit, where some incomplete changes are done on event types when the content block types it depends on are changed too.













<!-- Replace -->
----
Preview: https://pr-1440--ruma-docs.surge.sh
<!-- Replace -->

zecakeh committed:
events: Update types according to changes in MSCs 1767, 3954, 3955 and 3956

zecakeh committed:
events: Update types according to changes in MSC3551

zecakeh committed:
events: Update types according to changes in MSC3552

zecakeh committed:
events: Update types according to changes in MSC3553

zecakeh committed:
events: Update types according to MSC3927 and changes in MSC3246

zecakeh committed:
events: Update types according to changes in MSC3245

zecakeh committed:
events: Update types according to changes in MSC3554

zecakeh committed:
events: Update types according to changes in MSC3381

zecakeh commented:
There have been some minor changes again (mainly splitting some stuff in separate MSCs), I'll put this as a draft until I have time to update it.

zecakeh commented:
The "root" MSC1767 is now in [PFCP](https://github.com/matrix-org/matrix-spec-proposals/pull/1767#issuecomment-1382598157) so I'll wait until it passes FCP to update this, so we don't need to update it again.

turt2live commented:
Note that `m.markup` has been renamed to `m.text` as part of the conditions for entering FCP.

zecakeh committed:
events: Updates location event types

Even if the MSC hasn't been updated, refactor a bit to match other
extensible events types.

zecakeh commented:
@ruma/reviewers This is ready for review now.

turt2live commented:
err, just for clarity's sake: *just* MSC1767 has been accepted. The other MSCs are still considered in review and subject to change.

jplatte commented:
This is *a lot* 😅
I've reviewed the first commit though, that one seems pretty straight-forward. Feel free to push that one to `main` already.

zecakeh commented:
> Feel free to push that one to main already.

Done

jplatte commented on crates/ruma-common/src/events/poll/end.rs:
```diff
@@ -1,42 +1,95 @@
+
+impl FromIterator<(String, UInt)> for PollResultsContentBlock {
+    fn from_iter<T: IntoIterator<Item = (String, UInt)>>(iter: T) -> Self {
+        Self(iter.into_iter().collect())
```
```suggestion
        Self(BTreeMap::from_iter(iter))
```

jplatte replied:
If there is a `FromIterator` impl, a corresponding `IntoIterator` would be good too, right?

jplatte commented on crates/ruma-common/src/events/poll/response.rs:
```diff
@@ -22,27 +33,44 @@ pub struct PollResponseEventContent {
+
+impl FromIterator<String> for SelectionsContentBlock {
+    fn from_iter<T: IntoIterator<Item = String>>(iter: T) -> Self {
+        Self(iter.into_iter().collect())
```
Same here.
